server/src/leaderboard.ts

```diff
@@ -0,0 +1,328 @@
+          ROW_NUMBER() OVER (ORDER BY total_time) AS rank
+      FROM 
+          cumulative_times
+  ),
+  leaderboard AS (
+    SELECT 
+        'Level 1' AS category,
+        problem_id,
+        fastest_time AS time
+    FROM 
+        ranked_fastest_times
+    WHERE 
+        level = 1 AND rank <= 3
+    UNION ALL
+    SELECT 
+        'Level 2' AS category,
+        problem_id,
+        fastest_time AS time
+    FROM 
+        ranked_fastest_times
+    WHERE 
+        level = 2 AND rank <= 3
+    UNION ALL
+    SELECT 
+        'Level 3' AS category,
+        problem_id,
+        fastest_time AS time
+    FROM 
+        ranked_fastest_times
+    WHERE 
+        level = 3 AND rank <= 3
+    UNION ALL
+    SELECT 
+        'Cumulative' AS category,
+        problem_id,
+        total_time AS time
+    FROM 
+        ranked_cumulative_times
+    WHERE 
+        rank <= 3
+    ORDER BY 
+        category, time
+  ),
+  latest_transactions AS (
+    SELECT
+        token_id,
+        "to",
+        ROW_NUMBER() OVER (PARTITION BY token_id ORDER BY block_num DESC, tx_idx DESC, log_idx DESC) AS rn
+    FROM
+      problems_transfer
+    WHERE
+      token_id IN (SELECT problem_id FROM leaderboard)
+      AND src_name = '${chain}'
+  ),
+  current_owners AS (
+      SELECT
+          token_id,
+          concat('0x', encode("to", 'hex')) as owner
+      FROM
+          latest_transactions
+      WHERE
+          rn = 1
+  )
+  SELECT 
+      *
+  FROM
+    leaderboard
+  LEFT JOIN current_owners ON leaderboard.problem_id = current_owners.token_id
+  `)
+
+  function scores(rows: any[]): SpeedScore[] {
+    return rows.map((r: any) => ({
+      problemId: r.problem_id,
+      ticks: r.time,
+      owner: r.owner
+    }))
+  }
+
+  return {
+    oneBody: scores(result.rows.filter((r: any) => r.category === 'Level 1')),
+    twoBody: scores(result.rows.filter((r: any) => r.category === 'Level 2')),
+    threeBody: scores(result.rows.filter((r: any) => r.category === 'Level 3')),
+    cumulative: scores(
+      result.rows.filter((r: any) => r.category === 'Cumulative')
+    )
+  }
+}
+
+async function calculateAllTimeLeaderboard(
+  today: number,
+  chain: Chain
+): Promise<Leaderboard['allTime']> {
+  const n = 10
+  const result = await db.query(`
+  WITH most_solved AS (
+    SELECT 
+        problem_id,
+        COUNT(*) AS solve_count
+    FROM 
+        solver_solved
+    WHERE
+        src_name = '${chain}'
+    GROUP BY 
+        problem_id
+    ORDER BY 
+        solve_count DESC
+    LIMIT ${n}
+),
+current_streaks AS (
+    SELECT
+        problem_id,
+        MAX(streak_length) AS current_streak
+    FROM (
+        SELECT
+            problem_id,
+            COUNT(*) AS streak_length
+        FROM (
+            SELECT
+                problem_id,
+                day,
+                ROW_NUMBER() OVER (PARTITION BY problem_id ORDER BY day) - 
+                ROW_NUMBER() OVER (PARTITION BY problem_id, day ORDER BY day) AS streak
```

Comment:
this spookiness gets consecutive days?

Reply:
yeah it's fucked to look at. these are called window functions https://www.postgresql.org/docs/current/tutorial-window.html

try pasting the whole thing to gpt if you want a more socratic explanation :)